Symantec DLP v2

barchen1 opened the pull request:
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/25659

## Description
Symantec DLP v2

## Screenshots
![Screen Shot 2022-03-14 at 12 47 47](https://user-images.githubusercontent.com/54398957/158173433-f1b74854-9f91-4f36-a295-248603092d86.png)
![Screen Shot 2022-03-14 at 14 35 11](https://user-images.githubusercontent.com/54398957/158173452-282c74f3-f799-495b-80ba-e27921d593e6.png)
![Screen Shot 2022-03-14 at 14 35 36](https://user-images.githubusercontent.com/54398957/158173455-b0777a36-e145-4bc3-bbbe-b8aace34e009.png)
![Screen Shot 2022-03-14 at 14 37 49](https://user-images.githubusercontent.com/54398957/158173801-dd53c91a-0e6e-4931-a6ab-93dd51890c95.png)



## Minimum version of Cortex XSOAR
- [x] 6.0.0

## Must have
- [x] Tests
- [x] Documentation

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+
+def build_custom_attributes_update(custom_attributes: List[str]):
+    """
+    Builds the custom_attributes_list that the user wants to update.
```
let's enhance the docstring here

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+
+def get_incident_history_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    incident_id = arg_to_number(args.get('incident_id'))
+    limit = arg_to_number(args.get('limit', 50))
```
How big can the number of results be, in theory?

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+    if last_run and last_run.get('last_incident_creation_time'):
+        last_update_time = last_run.get('last_incident_creation_time')
+        last_incident_id = last_run.get('last_incident_id', '')
+
```
No space between the if and the else please, it's one block

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+    for severity_name, severity_num in INCIDENT_SEVERITY_MAPPING.items():
+        if severity_num == severity:
+            return severity_name
+    return ''
```
Can that final return statement ever happen or is it unreachable code?

guykeller replied:
We said we'll remove this

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+        incident_severities = [INCIDENT_SEVERITY_MAPPING[severity] for severity in incident_severities]  # type: ignore
+    if incident_types:
+        incident_types = [INCIDENT_TYPE_MAPPING[incident_type] for incident_type in incident_types]
+    incidents_data_list = client.get_incidents_request(creation_date=last_update_time, status_id=incident_status_id,
```
Let's improve the variable's name

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+        last_incident_id = ''
+
+    incidents = []
+    if incident_severities:
+        incident_severities = [INCIDENT_SEVERITY_MAPPING[severity] for severity in incident_severities]  # type: ignore
+    if incident_types:
+        incident_types = [INCIDENT_TYPE_MAPPING[incident_type] for incident_type in incident_types]
```
let's add an empty line after each if block

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+                                                       severity=incident_severities,  # type: ignore
+                                                       incident_type=incident_types, limit=fetch_limit, order_by="ASC")
+    incidents_data_list = incidents_data_list.get('incidents', [])
+    if incidents_data_list:
```
Why is this line necessary? can `incidents_data_list` be `None` or just empty?

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+            incident_id = incident_data.get('incidentId')
+            incident_creation_time = (dateparser.parse(incident_data.get('creationDate', ''))).strftime(
+                '%Y-%m-%dT%H:%M:%SZ')
+            if is_incident_already_fetched_in_previous_fetch(last_update_time, incident_creation_time, last_incident_id,
```
When and why would that happen?

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+def fetch_incidents(client: Client, fetch_time: str, fetch_limit: int, last_run: dict, incident_types: List[str] = None,
+                    incident_status_id: List[str] = None, incident_severities: List[str] = None):
+    """
+    Performs the fetch incidents functionality, which means that every minute if fetches incidents
+    from Symantec DLP and uploads them to Cortex XSOAR server.
```
Add a more detailed explanation about how the fetch works and why.
For example, why we might have already fetched incidents found more than once.
Non trivial knowledge that can't be reflected in the code should be recorded in the docstring

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+
+def get_incident_details_fetch(client, incident):
+    """
+    Create incident details dict for each incident pulled from the fetch
```
Document the 401 issue in the docstring with a detailed explanation

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+    try:
+        incident_id = incident.get('incidentId')
+        incident_type = incident.get('messageSource', '')
+        static_attributes = client.get_incident_static_attributes_request(incident_id)
+        editable_attributes = client.get_incident_editable_attributes_request(incident_id)
+        incident_details = get_common_incident_details(static_attributes, editable_attributes,
+                                                       args={"custom_attributes": "all"}, isfetch=True,
+                                                       incident_type=incident_type)
```
We see these 3 lines in more than one location, let's convert them to an external function called `get_full_incident_details` or some better name, with a suitable docstring

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+                                                       incident_type=incident_type)
+    # In case of getting 401 (Unauthorized incident) - will get missing data
+    except DemistoException as e:
+        if '401' in str(e):
+            incident_details = get_details_unauthorized_incident(incident)
```
else?

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+
+
+def test_module(client: Client) -> str:
+    """Tests API connectivity and authentication'
+
+    Returning 'ok' indicates that the integration works like it is supposed to.
+    Connection to the service is successful.
+    Raises exceptions if something goes wrong.
+
+    :type client: ``Client``
+    :param Client: client to use
+
+    :return: 'ok' if test passed, anything else will fail the test.
+    :rtype: ``str``
+    """
+    """
+    Performs basic get request to get item samples
+    """
```
1. remove the docstring please
2. Let's make sure we also check the validity of params in the test module

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+                'occurred': incident_creation_time
+            }
+            incidents.append(incident)
+            if incident_id == incidents_data_list[-1].get('incidentId'):
+                last_update_time = incident_creation_time
+                last_incident_id = incident_id
```
Why do we need both the time and the ID if we only compare the ID?
Let's discuss this logic as a whole flow

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml:
```diff
@@ -0,0 +1,417 @@
+  - Endpoint
+  required: false
+  type: 16
+- additionalinfo: The status ID of the incidents. In order to get status IDs, you
+    should run the command `symantec-dlp-list-incident-status`
```
let's discuss

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml:
```diff
@@ -0,0 +1,417 @@
+  name: incidentStatusId
+  required: false
+  type: 16
+- display: Incident Severity
+  hidden: false
+  name: incidentSeverity
+  options:
+  - Info
+  - Low
+  - Medium
+  - High
```
No `additionalinfo`?

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml:
```diff
@@ -0,0 +1,417 @@
+  commands:
+  - arguments:
+    - default: false
+      description: Get incidents with creation date later than specified. Support
```
```suggestion
      description: The earliest time from which to get incidents. Supports
```

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py:
```diff
@@ -0,0 +1,522 @@
+
+
+@freeze_time("2022-03-07T13:34:14Z")
+def test_fetch_incidents__single(requests_mock):
```
Why the double underscore in the name, here and everywhere?

guykeller replied:
change it please

guykeller commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py:
```diff
@@ -0,0 +1,522 @@
+    assert incidents[0].get('rawJSON') == SECOND_INCIDENT_DETAILS
+
+
+def test_fetch_incidents__last_run_no_fetch(requests_mock):
```
Let's discuss the test cases for the fetch mechanism

barchen1 commented:
@jrosenpanw Can you please go over the docs files?
Thanks!

richardbluestone commented:
I've approved the docs. I couldn't edit this one (only view) but it looks ok.
![image](https://user-images.githubusercontent.com/53567272/161015047-af44fa40-8353-4e08-a40d-4efd7f8d9fca.png)

barchen1 commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,909 @@
+    :param incident_severities: The incident severities to filter.
+    :return: A list of Cortex XSOAR incidents
+    """
+    if last_run and last_run.get('last_incident_creation_time'):
```
no, changed

barchen1 commented:
@guykeller @meital-aizen I changed the fetch-incidents function because of the issue we talked about. (Incident's ID and creation date are not synchronised)
I added a note about it in the README file as well.
Let me know if there is anything you would like me to add/change.

barchen1 commented on Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:
```diff
@@ -0,0 +1,923 @@
+    :param incident_creation_date: The current incident creation date
+
+    """
+    return last_update_time and last_update_time >= incident_creation_date
```
our assumption is that incidents are not created at the exact time (even if it's one millisecond difference).
@meital-aizen FYI

barchen1 commented:
@yuvalbenshalom We will need a force merge here - error on SDK-validate
![Screen Shot 2022-04-10 at 12 08 01](https://user-images.githubusercontent.com/54398957/162611073-7140903b-11c9-4772-b3f7-e293bcf242fd.png)

FYI @dorschw
